addons/crm/tests/test_crm_lead_multicompany.py

```diff
@@ -90,32 +90,81 @@ def test_lead_mc_company_computation_env_team_norestrict(self):
     @users('user_sales_manager_mc')
     def test_lead_mc_company_computation_env_user_restrict(self):
         """ Check that the computed company is allowed (aka in self.env.companies).
-        User is logged in company_main even his default default company is
-        company_2. """
+        User is logged in company_main even their default company is company_2.
+        Conversely, if no company is set on team, preferentially choose env current company."""
         LeadUnsyncCids = self.env['crm.lead'].with_context(allowed_company_ids=[self.company_main.id])
         self.assertEqual(LeadUnsyncCids.env.company, self.company_main)
         self.assertEqual(LeadUnsyncCids.env.companies, self.company_main)
         self.assertEqual(LeadUnsyncCids.env.user.company_id, self.company_2)
 
         # simulate auto-creation through sudo (assignment-like)
-        lead = LeadUnsyncCids.sudo().create({
-            'name': 'My Lead MC',
+        lead_1_auto = LeadUnsyncCids.sudo().create({
+            'name': 'My Lead MC 1 Auto',
         })
-        self.assertFalse(lead.company_id,
-                         'Lead: due to MC rule, avoid setting a company when it would cause crashes')
-        self.assertEqual(lead.team_id, self.sales_team_1,
-                         'Lead: due to MC rule, took first availability in other company')
-        self.assertEqual(lead.user_id, self.user_sales_manager_mc)
+        self.assertEqual(lead_1_auto.company_id, self.company_main)
```

Comment:
Would add a small msg to explain situation / expected value if this test fails (like: environment company is preferred over user's company when being in allowed companies if no team is involved, to avoid ACLs crash (if I understood everything, otherwise update message accordingly :D ))